File: 백준/Gold/7569. 토마토/토마토.py
import sys
input = sys.stdin.readline

# M : 가로, N : 세로, H : 높이
M, N, H = map(int, input().split())

box=[]
tomato = []
import queue

q = queue.Queue()
for i in range(H):
    box=[]
    for _ in range(N):
        row = list(map(int, input().split()))
        for j, r in enumerate(row):
            if r == 1:
                q.put([i, _, j])
        box.append(row)
    tomato.append(box)

# 우, 좌, 뒤, 앞, 위, 아래
dx = [0, 0, -1, 1, 0, 0]
dy = [1, -1, 0, 0, 0, 0]
dh = [0, 0, 0, 0, 1, -1]

day = 0
def bfs():
    global day

    while q.qsize():    
        for _ in range(q.qsize()):    
            pos = q.get()
            h = pos[0]
            x = pos[1]
            y = pos[2]

            # 우, 좌, 뒤, 앞, 위, 아래
            for i in range(6):
                if tomato[h][x][y] == -1:
                    break
                if x+dx[i] == -1 or y+dy[i] == -1 or h+dh[i] == -1 or x+dx[i] == N or y+dy[i] == M or h+dh[i] == H:
                    continue
                if tomato[h+dh[i]][x+dx[i]][y+dy[i]] == 0:
                    tomato[h+dh[i]][x+dx[i]][y+dy[i]] = 1
                    q.put([h+dh[i], x+dx[i], y+dy[i]])
        day+=1

bfs()

isfail = 0
for t in tomato:
    for i in t:
        for j in i:
            if j == 0:
                isfail = 1
                break
        if isfail:
            break
    if isfail:
        break

if isfail:
    print(-1)
else:
    print(day-1)

File: C++17 (Clang)/백준/Gold/7576. 토마토/토마토.cc
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

int M, N;
int dx[4] = {0, 0, -1, 1};
int dy[4] = {1, -1, 0, 0};

int main() {
    // 입력 받기
    cin >> M >> N;
    vector<vector<int>> tomato(N, vector<int>(M));
    queue<pair<int, int>> q;

	for (int n = 0; n < N; n++) {
		for (int m = 0; m < M; m++) {
			cin >> tomato[n][m];
			if (tomato[n][m] == 1) {
				q.push({n, m});
			}
		}
	}

    int day = 0;

    // BFS 시작
    while (!q.empty()) {
        int size = q.size();
        for (int i = 0; i < size; i++) {
            auto [x, y] = q.front();
            q.pop();

            for (int j = 0; j < 4; j++) {
                int nx = x + dx[j];
                int ny = y + dy[j];

                // 범위 검사
                if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;

                if (tomato[nx][ny] == 0) {
                    tomato[nx][ny] = 1;
                    q.push({nx, ny});
                }
            }
        }
        day++;
    }

    // 익지 않은 토마토가 남아있는지 검사
	for (int n = 0; n < N; n++) {
		for (int m = 0; m < M; m++) {
			if (tomato[n][m] == 0) {
				cout << -1 << endl;
				return 0;
			}
		}
	}

    // 최소 일수 출력
    cout << day - 1 << endl;

    return 0;
}